### PR TITLE
Rework Order Table

### PR DESCRIPTION
**Pull request by duncanspumpkin**
I said I would eventually get around to doing this so might as well start showing what I have.

Still need to create an iterator for this and some nice simple interaction code.

### CODE REVIEW

File: src/OpenLoco/Orders.h [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    };
+
+    struct OrderRouteWaypoint : Order
+    {
+        static constexpr OrderType TYPE = OrderType::RouteWaypoint;
+        uint8_t _1;
+        uint8_t _2;
+        uint8_t _3;
+        uint8_t _4;
+        uint8_t _5;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Orders.h:72]: Does the naming vary by waypoint? I'm assuming these are placeholders. If not, think we should name these `var_0X` at least.

Comment (duncanspumpkin) [reply src/OpenLoco/Orders.h:72]: I was copying the style from the tile elements. None of these fields should be accessed directly.

Comment (AaronVanGeffen) [reply src/OpenLoco/Orders.h:72]: Right, but surely they have some semantics? Like X/Y/Z tile coordinates and such?

**Comment by duncanspumpkin** [comment]
Note to self. Can't use unique ptr for clone function due to undefined behaviour without a virtual destructor. Use a shared ptr as we can't add a vtable.

File: src/OpenLoco/Windows/Vehicle.cpp [hunk 2]
```diff
@@ -2770,33 +2669,40 @@ namespace OpenLoco::Ui::Vehicle
                 return;
             }
 
-            switch (orderType)
+            if (selectedOrder == nullptr)
+            {
+                return;
+            }
+            switch (selectedOrder->getType())
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Windows/Vehicle.cpp:2676]: Not sure if you'd seen this yet, but:
```
../src/OpenLoco/Windows/Vehicle.cpp: In function 'void OpenLoco::Ui::Vehicle::Route::scrollMouseDown(OpenLoco::Ui::window*, int16_t, int16_t, uint8_t)':
../src/OpenLoco/Windows/Vehicle.cpp:2676:20: error: enumeration value 'UnloadAll' not handled in switch [-Werror=switch]
             switch (selectedOrder->getType())
                    ^
../src/OpenLoco/Windows/Vehicle.cpp:2676:20: error: enumeration value 'WaitFor' not handled in switch [-Werror=switch]
```

Comment (duncanspumpkin) [reply src/OpenLoco/Windows/Vehicle.cpp:2676]: Urgh yeah

File: src/OpenLoco/Orders.cpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    uint32_t Order::getOffset() const
+    {
+        return this - _orderTable;
+    }
+
+    uint64_t Order::getRaw() const
+    {
+        uint64_t ret = 0;
+        ret = _type;
+        switch (getType())
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Orders.cpp:119]:
```
 ../src/OpenLoco/Orders.cpp: In member function 'uint64_t OpenLoco::Vehicle::Order::getRaw() const':
../src/OpenLoco/Orders.cpp:119:16: error: enumeration value 'End' not handled in switch [-Werror=switch]
         switch (getType())
                ^
../src/OpenLoco/Orders.cpp:119:16: error: enumeration value 'UnloadAll' not handled in switch [-Werror=switch]
../src/OpenLoco/Orders.cpp:119:16: error: enumeration value 'WaitFor' not handled in switch [-Werror=switch]
```

**Comment by AaronVanGeffen** [comment]
The 'end of route list' item is no longer being drawn as of this PR:

master:
Screenshot (3)

This PR:
Screenshot (2)

Save file: [Let's Play Chris Sawyer's Locomotion (Randomland) by matsku84 - 2088.SV5.zip](https://github.com/OpenLoco/OpenLoco/files/5913989/Let.s.Play.Chris.Sawyer.s.Locomotion.Randomland.by.matsku84.-.2088.SV5.zip)

**Comment by duncanspumpkin** [comment]
I can't reproduce that on my saves but I'll try out that one as well.

**Comment by duncanspumpkin** [comment]
Think thats all of the outstanding issues with this corrected.

**Review by AaronVanGeffen — CHANGES_REQUESTED** [review]
- [ ] The 'end of list' item is still not accounted for when computing the scrollview height. This leads to it showing up when the window is resized, but not when it overflows.

- [ ] Regardless of whether the 'end of list' item is in view, I can't select it. I can do this on master.

**Comment by duncanspumpkin** [comment]
Argh what a pain I've made my iterators so that you don't have the end of list its whats causing these issues.

**Comment by duncanspumpkin** [comment]
Do we actually want to be able to select it? I don't see there being any reason why we would want to select it?

**Comment by AaronVanGeffen** [comment]
> Do we actually want to be able to select it? I don't see there being any reason why we would want to select it?

As I understand it, new items are added at the selected index, aren't they?

**Comment by duncanspumpkin** [comment]
Yes and if nothing is selected it goes at the end.

**Comment by AaronVanGeffen** [comment]
As you can't undo a selection, there should be a way to opt-in to that though, right?

**Comment by duncanspumpkin** [comment]
Okay I think it is now working how you want. I'm pretty sure it was only a display issue and not anything else

**Review by AaronVanGeffen — APPROVED** [review]

**Pull request merged**